jax_md/elasticity.py

```diff
@@ -0,0 +1,526 @@
+         [[[cxxxy, cxyxy], [cxyxy, cyyxy]],
+          [[cxxyy, cyyxy], [cyyxy, cyyyy]]]])
+
+  elif response_all.shape[0] == 21: 
+    cxxxx = response_all[0];
+    cyyyy = response_all[1];
+    czzzz = response_all[2];
+    cyzyz = response_all[3]/4.;
+    cxzxz = response_all[4]/4.;
+    cxyxy = response_all[5]/4.;
+    cyyzz = (response_all[6] - cyyyy - czzzz)/2.;
+    cxxzz = (response_all[7] - cxxxx - czzzz)/2.;
+    cxxyy = (response_all[8] - cxxxx - cyyyy)/2.;
+    cxxyz = (response_all[9] - cxxxx - 4.*cyzyz)/4.;
+    cxxxz = (response_all[10] - cxxxx - 4.*cxzxz)/4.;
+    cxxxy = (response_all[11] - cxxxx - 4.*cxyxy)/4.;
+    cyyyz = (response_all[12] - cyyyy - 4.*cyzyz)/4.;
+    cyyxz = (response_all[13] - cyyyy - 4.*cxzxz)/4.;
+    cyyxy = (response_all[14] - cyyyy - 4.*cxyxy)/4.;
+    czzyz = (response_all[15] - czzzz - 4.*cyzyz)/4.;
+    czzxz = (response_all[16] - czzzz - 4.*cxzxz)/4.;
+    czzxy = (response_all[17] - czzzz - 4.*cxyxy)/4.;
+    cyzxz = (response_all[18] - 4.*cyzyz - 4.*cxzxz)/8.;
+    cyzxy = (response_all[19] - 4.*cyzyz - 4.*cxyxy)/8.;
+    cxzxy = (response_all[20] - 4.*cxzxz - 4.*cxyxy)/8.;
+
+    C = jnp.array(
+         [[[[cxxxx, cxxxy, cxxxz],
+            [cxxxy, cxxyy, cxxyz],
+            [cxxxz, cxxyz, cxxzz]],
+           [[cxxxy, cxyxy, cxzxy],
+            [cxyxy, cyyxy, cyzxy],
+            [cxzxy, cyzxy, czzxy]],
+           [[cxxxz, cxzxy, cxzxz],
+            [cxzxy, cyyxz, cyzxz],
+            [cxzxz, cyzxz, czzxz]]],
+          [[[cxxxy, cxyxy, cxzxy],
+            [cxyxy, cyyxy, cyzxy],
+            [cxzxy, cyzxy, czzxy]],
+           [[cxxyy, cyyxy, cyyxz],
+            [cyyxy, cyyyy, cyyyz],
+            [cyyxz, cyyyz, cyyzz]],
+           [[cxxyz, cyzxy, cyzxz],
+            [cyzxy, cyyyz, cyzyz],
+            [cyzxz, cyzyz, czzyz]]],
+          [[[cxxxz, cxzxy, cxzxz],
+            [cxzxy, cyyxz, cyzxz],
+            [cxzxz, cyzxz, czzxz]],
+           [[cxxyz, cyzxy, cyzxz],
+            [cyzxy, cyyyz, cyzyz],
+            [cyzxz, cyzyz, czzyz]],
+           [[cxxzz, czzxy, czzxz],
+            [czzxy, cyyzz, czzyz],
+            [czzxz, czzyz, czzzz]]]])
+  else:
+    raise AssertionError('incorrect shape for array of responses')
+  return C
+
+def AthermalElasticModulusTensor(energy_fn: Callable[..., Array], 
+                                 tether_strength: float=1e-10,
+                                 gradient_check: Array=None, 
+                                 cg_tol: float=1e-7,
+                                 check_convergence: bool=False
+                                 ) -> Callable[..., Array]:
+  """ Setup calculation of elastic modulus tensor for a 2d or 3d athermal system.
+
+  The elastic modulus tensor describes a material's response to different 
+  boundary deformations. Specifically, for small deformation given by a 
+  symmetric strain tensor e, the change in energy is
+    U / V^0 = U^0/V^0 + s^0_{ij} e_{ji} + (1/2) C_{ijkl} e_{ij} e_{kl} + ...
```

Comment:
Want to add a citation here? It could be to one of your papers?

Reply:
Hummmm.... I'm actually not sure what the most appropriate thing to cite would be. Let me think on it.